### PR TITLE
Add extension manager

### PR DESCRIPTION
**Pull request by vidartf**
Integrates the code from [jupyterlab-discovery](https://github.com/vidartf/jupyterlab_discovery) (with some cleanup).

### CODE REVIEW

**Comment by vidartf** [comment]
The CI failure is significant, and I see that locally as well. Locally I can work around it by manually building things in this order:
- coreutils
- observables
- services
- everything

Any ideas?

**Comment by vidartf** [comment]
The problem was that the code imported from `@jupyterlab/services/lib/builder`. Now it import from `@jupyterlab/services` after adding an export of `Builder` typings namespace in services.

Note: I also added the package to singleton packages, and some deps to vendored, but I just tried to cargo cult, and I have no idea if I did this right.

**Comment by vidartf** [comment]
Rebased on latest.

**Comment by afshin** [comment]
There are some issues we need to consider regarding this PR:

* You can **use the discovery UI** if you check out and build this branch. However, once you install an extension, you won't be able to **use the newly installed extension** because you will have been running ([future perfect progressive tense](https://en.wikipedia.org/wiki/Uses_of_English_verb_forms#Future_perfect_progressive)) off of `master` in `--dev-mode`.
* If you `pip install jupyterlab_discovery`, you can install this exact functionality into your released version of JupyterLab in order to see it working *and* use the extensions it installs.
* This means that if we want to be able to test this functionality in the wild, we'll need it to not simply reside in `master` but actually be in a release. This is what @vidartf and I discussed today, and we think that having this extension in the Beta 3 pre-release gives us the ability to get some eyes on the plugin and decide whether to disable it (probably using the setting system and a default value) for the actual Beta 3 release or not. Either way, though, we'd propose *shipping* this code, but potentially making it disabled (again, I'm using the word "disable" informally, I don't mean disabling via the extension disable mechanism, but by using a setting).
* One of the potential pitfalls is that a `labextension` that also has a `serverextension` component cannot be installed by this current version of the extension. Vidar has experimental code for this in the external version, but it was not ready to put into this PR. The UI of this version will give users a meaningful response if the extension you're trying to install has metadata indicating it has a server extension.

Because this is a big and highly requested feature, I suggest we merge this and put it in the Beta 3 pre-release and iterate on it. There are some UI improvements we've already started talking about (*e.g.*, disabling the "Install" button after it has been clicked, organizing the list of extensions into collapsible sections for those that are install and those that are available, *etc.*)

cc: @jasongrout @ian-r-rose @ellisonbg

**Comment by ian-r-rose** [comment]
I think this is a great enhancement. A couple of thoughts/questions:

* +1 on including in the prerelease.
* We have gone to some length to avoid a dependency on node to use the core packages. This would introduce such a dependency. I think that for this reason we should disable it by default, either with the settings system or with the build parameters.
* In a similar vein, I just tested what happens if you try to use it without node installed, and it does not give a very clear error message. I think we should make it very clear what the problem is if node is not installed.
* Is there any way to inspect from the `package.json` of the extension whether it is compatible with the current JupyterLab distribution? That is to say, can it fail before trying to install it?
* What error will be shown if the user doesn't have write permissions in `JUPYTERLAB_DIR`?

Thanks for pushing forward on this @vidartf!

**Comment by vidartf** [comment]
@ian-r-rose I agree that the extension need to handle missing node better. Good point!

> Is there any way to inspect from the package.json of the extension whether it is compatible with the current JupyterLab distribution?

This is not a cheap operation, so filtering a search to only compatible versions seem non-trivial. On the server this PR adds some code for checking for an update to any of the installed extensions, which does something similar to that. It could maybe be hinted at with keywords, but I'm not sure if that is a robust solution.

**Comment by vidartf** [comment]
> What error will be shown if the user doesn't have write permissions in `JUPYTERLAB_DIR`

I haven't tried, but hopefully the same error as the user sees in the terminal (if not, we should ensure that this happens). Whether that error message is currently helpful is another issue.

**Comment by jasongrout** [comment]
We had a discussion this morning triaging issues for Beta 3. We thought we should include this PR, but disable the plugin by default so that the stock pip installed jupyterlab does not depend on nodejs. @afshin mentioned writing a setting that disables this plugin by default, but allows the plugin to be enabled.

**Comment by vidartf** [comment]
@afshin Do you have any recommendation for where to put the enable/disable code, and an outline for how it should look? I've not been tracking the settings system that closely. Also, if this is going to be disabled by default, I guess it makes sense to only have the server extension check for updates on request (now it checks on load for efficiency).

**Comment by afshin** [comment]
Here is how the enabled/disabled status works:

extension-manager-enabled

File: packages/extensionmanager-extension/src/index.ts [hunk 1]
```diff
@@ -33,25 +35,36 @@ namespace CommandIDs {
...
-  activate: (app: JupyterLab, restorer: ILayoutRestorer) => {
+  requires: [ILayoutRestorer, ISettingRegistry, IRouter],
+  activate: async (app: JupyterLab, restorer: ILayoutRestorer, registry: ISettingRegistry, router: IRouter) => {
+    const settings = await registry.load(plugin.id);
+    const enabled = settings.composite['enabled'] as boolean;
+
+    // If the extension is enabled or disabled, refresh the page.
+    settings.changed.connect(() => { router.reload(); });
+
+    if (!enabled) {
```
Comment (vidartf) [inline packages/extensionmanager-extension/src/index.ts]: Maybe `!== true` to guard against other falsy expressions?

File: packages/extensionmanager-extension/schema/plugin.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+  "jupyter.lab.setting-icon-class": "jp-SettingsIcon",
+  "jupyter.lab.setting-icon-label": "Extension Manager",
+  "title": "Extension Manager",
+  "description": "Extension manager settings.",
+  "properties": {
+    "enabled": {
+      "title": "Enabled Status",
+      "description": "Whether the extension manager is enabled.",
```
Comment (jasongrout) [inline packages/extensionmanager-extension/schema/plugin.json]: This should say something about requiring node, I think.

File: packages/extensionmanager-extension/src/index.ts (hunk 1 above)
Comment (jasongrout) [reply packages/extensionmanager-extension/src/index.ts]: Should a menu item say something about `(requires node)`? Or even better, is there some nice error if they don't have node installed? I think we talked about this in our triage - even better would be a good error if a rebuild isn't possible.

I hesitate to make the setting too discoverable until we have a good clear way for a user that doesn't have node installed to know why things don't work.

**Comment by afshin** [comment]
@jasongrout What do you think of this:

screen shot 2018-06-29 at 15 19 36

**Comment by afshin** [comment]
Ah okay. There was a version mismatch, 2.9.1 vs. 2.9.2. I'll run the integrity utility again.

**Comment by afshin** [comment]
Less horizontal real estate:

screen shot 2018-06-29 at 15 41 28

**Pull request merged**